File: src/services/chord/ChordDetectionService.ts
/**
 * ChordDetectionService
 * --------------------------------------------------
 * Real-time chord detection service (v1).
 *
 * Pipeline:
 *   Audio samples
 *     → Chroma vector
 *     → Template matching
 *     → Best chord decision
 *     → ChordCandidate | null
 *
 * NOTE:
 * This service DOES NOT decide whether a chord is
 * musically meaningful. It only produces raw analysis
 * results. Semantic decisions are handled by AudioPipeline.
 */

import {
    ChordDetectionService as IChordDetectionService,
    ChordDetectionResult,
} from './ChordDetectionTypes';

import { computeChromaVector } from './Chroma';
import { generateAllResolvedTemplates } from './ChordTemplates';
import { findBestChordCandidate } from './utils/decision';

/* -------------------------------------------------- */
/* Constants                                          */
/* -------------------------------------------------- */

const DEFAULT_SAMPLE_RATE = 44100;

/* -------------------------------------------------- */
/* Service Implementation                             */
/* -------------------------------------------------- */

export class ChordDetectionService
    implements IChordDetectionService
{
    /* ---------------------------------------------- */
    /* Internal State                                 */
    /* ---------------------------------------------- */

    private sampleRate: number = DEFAULT_SAMPLE_RATE;
    private running: boolean = false;

    private callback:
        | ((result: ChordDetectionResult | null) => void)
        | null = null;

    /**
     * Pre-generated chord templates (C major, C minor, ... B minor)
     * Generated once and reused for performance.
     */
    private readonly templates = generateAllResolvedTemplates();

    /* ---------------------------------------------- */
    /* Lifecycle                                      */
    /* ---------------------------------------------- */

    start(
        callback: (result: ChordDetectionResult | null) => void
    ): void {
        if (this.running) return;

        this.running = true;
        this.callback = callback;
    }

    stop(): void {
        if (!this.running) return;

        this.running = false;
        this.callback = null;
    }

    setSampleRate(sampleRate: number): void {
        this.sampleRate = sampleRate;
    }

    /* ---------------------------------------------- */
    /* Main Processing                                */
    /* ---------------------------------------------- */

    processSamples(
        samples: Float32Array,
        timestamp: number
    ): void {
        if (!this.running || !this.callback) return;
        if (!samples || samples.length === 0) return;

        /* FFT → Chroma (chroma vector oluşturuluyor) */
        const chroma = computeChromaVector(
            samples,
            this.sampleRate
        );

        /* Total chroma energy (used by pipeline for gating) */
        const chromaEnergy = chroma.reduce(
            (sum, v) => sum + v,
            0
        );

        /* Template matching → Best chord (cosine sim. + decision.ts) */
        const candidate = findBestChordCandidate(
            chroma,
            this.templates
        );

        /* Emit raw analysis result */
        const result: ChordDetectionResult = {
            timestamp,
            candidate,
            chroma,
            chromaEnergy,
        };

        this.callback(result);
    }
}

File: ios/ChordDetection/ChordDetectionService.swift
//
//  ChordDetectionService.swift
//  SmartTabGuitarKaraoke
//
//  JS parity implementation of ChordDetectionService
//

import Foundation

// MARK: - Constants

private let DEFAULT_SAMPLE_RATE: Double = 44100.0

// MARK: - Service

final class ChordDetectionService {

    // MARK: - Internal State

    private var sampleRate: Double = DEFAULT_SAMPLE_RATE
    private var running: Bool = false

    private var callback: ((ChordDetectionResult) -> Void)?

    /// Pre-generated chord templates (C major, C minor, ... B minor)
    /// Generated once and reused for performance.
    private let templates: [ResolvedChordTemplate] =
        generateAllResolvedTemplates()

    // MARK: - Lifecycle

    func start(
        callback: @escaping (ChordDetectionResult) -> Void
    ) {
        if running { return }

        running = true
        self.callback = callback
    }

    func stop() {
        if !running { return }

        running = false
        callback = nil
    }

    func setSampleRate(_ sampleRate: Double) {
        self.sampleRate = sampleRate
    }

    // MARK: - Main Processing (JS parity)

    func processSamples(
        samples: [Float],
        timestampMs: Int
    ) {
        // JS parity guards
        if !running { return }
        guard let cb = callback else { return }
        if samples.isEmpty { return }

        // FFT → Chroma
        let chroma = computeChromaVector(
            samples: samples,
            sampleRate: sampleRate
        )

        // Total chroma energy
        let chromaEnergy = chroma.reduce(0.0, +)

        // Template matching → Best chord
        let candidate = findBestChordCandidate(
            chroma: chroma,
            templates: templates
        )

        // Emit raw analysis result (NO semantics)
        let result = ChordDetectionResult(
            timestamp: timestampMs,
            candidate: candidate,
            chroma: chroma,
            chromaEnergy: chromaEnergy
        )

        cb(result)
    }
}
